Secure ROS modifications

asundaresan opened the pull request:
These are a series of commits that introduce Secure ROS features into ROS. The commits are broken up so that the features are introduced into rosmaster, rosgraph, roslaunch, rospy, xmlrpcpp, roscpp in roughly that order.

asundaresan committed:
Add test_secure_ros

asundaresan committed:
Add Secure ROS features to master

asundaresan committed:
Add Secure ROS features to rospy

asundaresan committed:
Add client information in xmlrpc method callback

asundaresan committed:
Add Secure ROS features to roscpp

asundaresan committed:
Add path resolution for ROS_AUTH_FILE

asundaresan committed:
Add secure XMLRPC method to roslaunch

dirk-thomas commented:
Thank you making the PR. Due to the size (~2.5k lines) and its invasiveness it will likely take a while to review it. In the meantime it would be good if you could look into the failing tests and address them as well as the new compiler warning.

asundaresan committed:
Fix unused parameter warning in XmlRpcServer.cpp

asundaresan commented:
Hi Dirk:

I can fix the GCC warning. However, I'm not sure about the tests. Could you
point out to me more documentation of how the tests work in ROS? How can I
replicate the jenkins build and test?  I get test failures on both the
*lunar-devel* branch of *ros_comm* and my own commits! I'd appreciate any
pointers.

Thanks,
Aravind.

On Thu, Jun 15, 2017 at 4:36 PM, Dirk Thomas <notifications@github.com>
wrote:

> Thank you making the PR. Due to the size (~2.5k lines) and its
> invasiveness it will likely take a while to review it. In the meantime it
> would be good if you could look into the failing tests and address them as
> well as the new compiler warning.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ros/ros_comm/pull/1080#issuecomment-308893624>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHwIS4XsgZ9OnwmjZQlmh6LxSjQojOkPks5sEcAMgaJpZM4N70KO>
> .
>

asundaresan committed:
Fix roslaunch test errors

asundaresan committed:
Allow local IP addresses for clients

asundaresan committed:
Fix client_ip_address tuple bug

asundaresan committed:
Add checks for auth.noverify

asundaresan committed:
Change default logging level to WARN

asundaresan committed:
Fix service clients for noverify=True

asundaresan committed:
Add client_ip_address to missed methods

asundaresan committed:
Client checks only if Secure ROS enabled

asundaresan commented:
Hi Dirk:
I've fixed the failing tests and rebased the PR on top of the most recent commit in lunar-devel. Please let me know if you have any questions about any of the commits in the PR. 
Thanks,
Aravind.

PeterMitrano commented on clients/roscpp/include/ros/topic_manager.h:
```diff
@@ -60,6 +64,9 @@ typedef boost::shared_ptr<ConnectionManager> ConnectionManagerPtr;
 class SubscriptionCallbackHelper;
 typedef boost::shared_ptr<SubscriptionCallbackHelper> SubscriptionCallbackHelperPtr;
 
+bool is_subscriber_authorized( const std::string& topic, const std::string& client_ip_address );
```
extra spaces here do not match the style of the surrounding code

PeterMitrano commented on clients/roscpp/include/ros/topic_manager.h:
```diff
@@ -210,7 +217,9 @@ class ROSCPP_DECL TopicManager
   bool pubUpdate(const std::string &topic, const std::vector<std::string> &pubs);
 
   void pubUpdateCallback(XmlRpc::XmlRpcValue& params, XmlRpc::XmlRpcValue& result);
+  void pubUpdateCallback(XmlRpc::XmlRpcValue& params, XmlRpc::XmlRpcValue& result, XmlRpc::XmlRpcClientInfo& client_info);
```
a general comment, seems like a lot of these parameters can be references to const, i.e. `XmlRpc::XmlRpcValue const &` or  `const XmlRpc::XmlRpcValue &`

PeterMitrano commented on clients/roscpp/src/libros/url.cpp:
```diff
@@ -0,0 +1,136 @@
+  this->parse(url_s);
+}
+
+URLParser::~URLParser()
```
why define a destructor that does nothing?

PeterMitrano commented on clients/roscpp/include/ros/url.h:
```diff
@@ -0,0 +1,30 @@
+
+    std::string get_host();
+
+    inline std::string get_ip_address() { return ip_address_; }
```
many of these member functions can be const

PeterMitrano commented on clients/roscpp/src/libros/url.cpp:
```diff
@@ -0,0 +1,136 @@
+            ptr = &((struct sockaddr_in6 *) res->ai_addr)->sin6_addr;
+            break;
+        }
+        inet_ntop (res->ai_family, ptr, addrstr, 100);
```
can we be sure these are the only two cases? I'm pretty sure it's an error to pass a uninitialized ptr to inet_ntop

PeterMitrano commented on test/test_secure_ros/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+#  DEPENDS system_lib
+)
+
+include_directories( ${catkin_INCLUDE_DIRS} )
```
consider target_include_directories to prevent "leaking" include directories to other targets

PeterMitrano commented on test/test_secure_ros/nodes/test_self.py:
```diff
@@ -0,0 +1,193 @@
+
+
+"""
+Read in an authorization file and test Secure ROS implementation by probing master.
```
would it make sense to write this as a "rostest"? one advantage is that it would be able to launch the nodes necessary to perform these tests

PeterMitrano commented on test/test_secure_ros/src/is_uri_match.cpp:
```diff
@@ -0,0 +1,55 @@
+
+int main(int argc, char *argv[])
+{
+	if (argc < 2) {
```
strange indentation in this file

PeterMitrano commented on utilities/xmlrpcpp/include/xmlrpcpp/XmlRpcServer.h:
```diff
@@ -107,7 +115,10 @@ namespace XmlRpc {
     int _port;
-
+    XmlRpcClientInfo _clientInfo;
+    std::map<int,std::string> map_ip;
```
could these be unordered_map? they are much faster

PeterMitrano commented on tools/rosmaster/src/rosmaster/authorization.py:
```diff
@@ -0,0 +1,482 @@
+    def save( self, filename ):
+        """ Write full authorization configuration to YAML file
+        """
+        data = OrderedDict()
```
if we're targeting Noetic that means python3, dicts are ordered by default so this is unnecessary. Also, why not just one call to update with one big dictionary?

Also still noticing lots of extra spaces around params and curly braces, not sure if this is a problem with the style of the existing code.

PeterMitrano reviewed (commented):
not a maintainer but I thought I'd leave a few minor comments just cuz I was curious

sloretz commented:
Thank you for the PR!

Unfortunately I don't think we should merge this one.

[ROS Noetic will reach end-of-life on May 31st, 2025](https://discourse.ros.org/t/ros-noetic-end-of-life-may-31-2025/43160). Every change comes with a risk of introducing regressions, and there isn't much time left to fix them. I'm closing pull requests that add features so that the remaining time is allocated towards bug fixes and compatibility with newer Ubuntu distros.
